### PR TITLE
[BEAM-9608] Increasing scope of context managers for FnApiRunner

### PR DESCRIPTION
**Pull request by pabloem**
**Please** add a meaningful description for your change here

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/)

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by pabloem** [comment]
r: @robertwb 
I've added two classes under `execution`: 
- `FnApiExecutionContext`, which manages runner-related context utilities and variables, like intermediate pcollection buffers, worker handler manager, safe coders and pipeline context.
- `BundleContextManager`, which manages bundle-exeecution-related utilities and variables, such as the process bundle descriptor, the worker handlers for that stage, and the callbacks for get_buffer and get_coder_impl.

This change increases their scope to manage iterable_state_write, and worker handlers. The goal in the future (streaming) is for `FnApiExecutionContext` to manage a watermark manager.

File: sdks/python/apache_beam/runners/portability/fn_api_runner/fn_runner.py [hunk 1]
```diff
@@ -369,8 +366,8 @@ def _store_side_inputs_in_state(self,
           state_key = beam_fn_api_pb2.StateKey(
               iterable_side_input=beam_fn_api_pb2.StateKey.IterableSideInput(
                   transform_id=transform_id, side_input_id=tag, window=window))
-          bundle_context_manager.worker_handler.state.append_raw(
-              state_key, elements_data)
+          runner_execution_context.worker_handler_manager.state_servicer\
```
Comment (robertwb) [inline sdks/python/apache_beam/runners/portability/fn_api_runner/fn_runner.py]: Perhaps add a state_servicer() method right on runner_execution_context (and use several places below)?

File: sdks/python/apache_beam/runners/portability/fn_api_runner/fn_runner.py [hunk 2]
```diff
@@ -415,24 +412,24 @@ def _run_bundle_multiple_times_for_testing(
...
-      stage,  # type: translations.Stage
+      runner_execution_context,  # type: execution.FnApiRunnerExecutionContext
       bundle_context_manager,  # type: execution.BundleContextManager
       deferred_inputs,  # type: MutableMapping[str, PartitionableBuffer]
-      data_channel_coders,  # type: Mapping[str, str]
   ):
     # type: (...) -> None
 
-    for transform_id, timer_writes in stage.timer_pcollections:
+    for transform_id, timer_writes in \
```
Comment (robertwb) [inline sdks/python/apache_beam/runners/portability/fn_api_runner/fn_runner.py]: backslash

File: sdks/python/apache_beam/runners/portability/fn_api_runner/fn_runner.py (hunk 1 above)
Comment (robertwb) [reply sdks/python/apache_beam/runners/portability/fn_api_runner/fn_runner.py]: On second thought, perhaps the ownership of state servicer should be moved up to runner_execution_context (though the worker manager may need a reference). Your call.

File: sdks/python/apache_beam/runners/portability/fn_api_runner/execution.py [hunk 3]
```diff
@@ -245,37 +254,106 @@ class FnApiRunnerExecutionContext(object):
...
   def __init__(self,
-      worker_handler_factory,  # type: Callable[[Optional[str], int], List[WorkerHandler]]
+      worker_handler_manager,  # type: worker_handlers.WorkerHandlerManager
       pipeline_components,  # type: beam_runner_api_pb2.Components
       safe_coders,
       data_channel_coders,
                ):
     """
-    :param worker_handler_factory: A ``callable`` that takes in an environment
+    :param worker_handler_manager: A ``callable`` that takes in an environment
```
[the remark this answers is not in the document]
Comment (pabloem) [reply sdks/python/apache_beam/runners/portability/fn_api_runner/execution.py]: Done.

File: sdks/python/apache_beam/runners/portability/fn_api_runner/execution.py [hunk 4]
```diff
@@ -245,37 +254,106 @@ class FnApiRunnerExecutionContext(object):
...
+    self.num_workers = num_workers
+
+    # Properties that are lazily initialized
+    self._process_bundle_descriptor = None
+    self._worker_handlers = None
+
+  @property
+  def worker_handlers(self):
+    if self._worker_handlers is None:
+      self._worker_handlers = self.execution_context.worker_handler_manager\
```
[the remark this answers is not in the document]
Comment (pabloem) [reply sdks/python/apache_beam/runners/portability/fn_api_runner/execution.py]: Done.

File: sdks/python/apache_beam/runners/portability/fn_api_runner/execution.py [hunk 5]
```diff
@@ -245,37 +254,106 @@ class FnApiRunnerExecutionContext(object):
...
+        .get_worker_handlers(self.stage.environment, self.num_workers)
+    return self._worker_handlers
+
+  def data_api_service_descriptor(self):
+    # All worker_handlers share the same grpc server, so we can read grpc server
+    # info from any worker_handler and read from the first worker_handler.
+    return next(iter(self.worker_handlers)).data_api_service_descriptor()
```
[the remark this answers is not in the document]
Comment (pabloem) [reply sdks/python/apache_beam/runners/portability/fn_api_runner/execution.py]: Done.

File: sdks/python/apache_beam/runners/portability/fn_api_runner/execution.py [hunk 6]
```diff
@@ -245,37 +254,106 @@ class FnApiRunnerExecutionContext(object):
...
+            self.execution_context.pipeline_components.coders.items()),
+        windowing_strategies=dict(
+            self.execution_context.pipeline_components.windowing_strategies.
+              items()),
+        environments=dict(
+            self.execution_context.pipeline_components.environments.items()))
+
+    state_api_serv_desc = self.state_api_service_descriptor()
+    if state_api_serv_desc:
+      res.state_api_service_descriptor.url = state_api_serv_desc.url
```
[the remark this answers is not in the document]
Comment (pabloem) [reply sdks/python/apache_beam/runners/portability/fn_api_runner/execution.py]: Done.

File: sdks/python/apache_beam/runners/portability/fn_api_runner/fn_runner.py (hunk 1 above)
Comment (pabloem) [reply sdks/python/apache_beam/runners/portability/fn_api_runner/fn_runner.py]: I've just added the state_sevicer() from your first comment, but I've created a jira issue to track fixng that later on.

File: sdks/python/apache_beam/runners/portability/fn_api_runner/fn_runner.py (hunk 2 above)
Comment (pabloem) [reply sdks/python/apache_beam/runners/portability/fn_api_runner/fn_runner.py]: Done

**Comment by pabloem** [comment]
@robertwb thoughts?

File: sdks/python/apache_beam/runners/portability/fn_api_runner/fn_runner.py (hunk 1 above)
Comment (robertwb) [reply sdks/python/apache_beam/runners/portability/fn_api_runner/fn_runner.py:369]: Link to JIRA for refernce?

File: sdks/python/apache_beam/runners/portability/fn_api_runner/fn_runner.py [hunk 7]
```diff
@@ -511,87 +509,32 @@ def _add_residuals_and_channel_splits_to_deferred_inputs(
 
   def _run_stage(self,
                  runner_execution_context,  # type: execution.FnApiRunnerExecutionContext
-                 stage,  # type: translations.Stage
+                 bundle_context_manager,  # type: execution.BundleContextManager
```
Comment (robertwb) [inline sdks/python/apache_beam/runners/portability/fn_api_runner/fn_runner.py:512]: It's odd that _run_stage no longer takes as a parameter the stage to run. Perhaps bundle_context_manager (and its class?) should be named stage_context or similar?

File: sdks/python/apache_beam/runners/portability/fn_api_runner/execution.py [hunk 8]
```diff
@@ -245,37 +254,108 @@ class FnApiRunnerExecutionContext(object):
...
+    # info from any worker_handler and read from the first worker_handler.
+    return self.worker_handlers[0].state_api_service_descriptor()
+
+  @property
+  def process_bundle_descriptor(self):
+    if self._process_bundle_descriptor is None:
+      self._process_bundle_descriptor = self._build_process_bundle_descriptor()
+    return self._process_bundle_descriptor
+
+  def _build_process_bundle_descriptor(self):
```
Comment (robertwb) [inline sdks/python/apache_beam/runners/portability/fn_api_runner/execution.py:342]: I was bitten by the fact that it is an error to access the process_bundle_descriptor before _extract_outputs is called. This should be clearly documented (and similarly for the other lazy attributes(s) in this class). Given that that's called external to this class and can't easily be checked, makes me wonder if the boundary of encapsulation needs to be adjusted here.

File: sdks/python/apache_beam/runners/portability/fn_api_runner/fn_runner.py (hunk 7 above)
Comment (robertwb) [reply sdks/python/apache_beam/runners/portability/fn_api_runner/fn_runner.py:512]: On this note, perhaps it makes sense to break FnApiRunner into the (mostly stateless) runner that can execute multiple pipelines and an executor (that has methods like run_stage) that might be stateful and is initialized with and tasked with running a single pipeline. Much of what is on context(s) would become state of self of this new object.

File: sdks/python/apache_beam/runners/portability/fn_api_runner/execution.py (hunk 8 above)
Comment (pabloem) [reply sdks/python/apache_beam/runners/portability/fn_api_runner/execution.py:342]: I was also bitten by this. I am trying to think of a better way to mark a Read transform as an Impulse Read, but can't think of one at the moment.

I'll add documentation for this. 

Once it's full streaming, the IMPULSE buffers will be enqued before the pipeline starts, and extract_outputs will work with non-impulse buffers only - thus we can modify the transform payload before starting the pipeline, and not compute the bundle descriptor lazily.

File: sdks/python/apache_beam/runners/portability/fn_api_runner/fn_runner.py (hunk 7 above)
Comment (pabloem) [reply sdks/python/apache_beam/runners/portability/fn_api_runner/fn_runner.py:512]: I'm adding all these comments to https://github.com/apache/beam/pull/11270

**Pull request merged**